### PR TITLE
8305746: InitializeEncoding should cache Charset object instead of charset name

### PR DESCRIPTION
**Pull request by peter-hofer**
Store `Charset` object in `jnuEncoding` and use `String(byte[], Charset)` and `String.getBytes(Charset)` instead of passing the charset name.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305746](https://bugs.openjdk.org/browse/JDK-8305746): InitializeEncoding should cache Charset object instead of charset name

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13499/head:pull/13499` \
`$ git checkout pull/13499`

Update a local copy of the PR: \
`$ git checkout pull/13499` \
`$ git pull https://git.openjdk.org/jdk.git pull/13499/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13499`

View PR using the GUI difftool: \
`$ git pr show -t 13499`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13499.diff">https://git.openjdk.org/jdk/pull/13499.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13499#issuecomment-1512894106)

### CODE REVIEW

**Comment by peter-hofer** [comment]
@naotoj please find here my changes to `InitializeEncoding`.

File: src/java.base/share/native/libjava/jni_util.c [hunk 1]
```diff
@@ -733,31 +730,38 @@ InitializeEncoding(JNIEnv *env, const char *encname)
...
-                // jnuEncoding falls back to UTF-8
-                jstring utf8 = (*env)->NewStringUTF(env, "UTF-8");
-                if (utf8 == NULL) {
-                    (*env)->DeleteLocalRef(env, enc);
-                    return;
-                }
+            if (!exc && charset.l != NULL) {
+                jnuEncoding = (*env)->NewGlobalRef(env, charset.l);
+                (*env)->DeleteLocalRef(env, charset.l);
+                break;
```
Comment (naotoj) [inline src/java.base/share/native/libjava/jni_util.c:757]: Could return immediately

File: src/java.base/share/native/libjava/jni_util.c [hunk 2]
```diff
@@ -625,11 +625,11 @@ getStringCp1252Chars(JNIEnv *env, jstring jstr)
 }
 
 static int fastEncoding = NO_ENCODING_YET;
-static jstring jnuEncoding = NULL;
+static jobject jnuEncoding = NULL;
```
Comment (naotoj) [inline src/java.base/share/native/libjava/jni_util.c:628]: Maybe `jnuCharset` is clearer that it is a `Charset` object.

**Review by naotoj — COMMENTED** [review]
Looks good to me (with minor nits). Since it would be hard to write a regression test, what kind of tests have you done? Did you run your patch on platforms that made sure your new code path work?

File: src/java.base/share/native/libjava/jni_util.c (hunk 1 above)
Comment (peter-hofer) [reply src/java.base/share/native/libjava/jni_util.c:757]: No, then the method id cache would not get initialized below. But I will add a comment.

Comment (naotoj) [reply src/java.base/share/native/libjava/jni_util.c]: You are right, those init code should follow.

**Review by naotoj — APPROVED** [review]
Looks good to me. Thanks for the change.

**Pull request closed without merging**